libraries/abstractions/ble_hal/test/src/iot_test_ble_hal_common.c

```diff
@@ -1765,47 +1773,134 @@ void prvRequestWriteCb( uint16_t usConnId,
 
-        if( ( pEventIndex->xEventTypes == xEventName ) &&
-            ( pEventIndex->lHandle == lhandle ) )
+        if( ( pEvent->xEventTypes == xEventName ) &&
+            ( pEvent->lHandle == lhandle ) )
         {
-            pvPtr = pEventIndex;
-            IotListDouble_Remove( &pEventIndex->eventList );
-
-            if( ( pxMatch != NULL ) && ( pvPtr != NULL ) && !pxMatch( pvPtr ) )
+            if( ( pxMatch == NULL ) || ( pxMatch( pEvent ) == true ) )
             {
-                pvPtr = NULL;
+                IotListDouble_Remove( &pEvent->eventList );
+                memcpy( pEventBuffer, pEvent, bufferLength );
+                IotTest_Free( pEvent );
+                xStatus = eBTStatusSuccess;
+                break; /* If the right event is received, exit. */
             }
+        }
+    }
+
+    IotMutex_Unlock( &threadSafetyMutex );
+
+    return xStatus;
+}
+
+static BLEHALEventsTypes_t checkQueueForEvents( BLEHALEventsTypes_t xEventsToCheck )
+{
+    BLEHALEventsTypes_t eventSet = eBLEHALEventNone;
+    BLEHALEventsInternals_t * pEvent;
+    IotLink_t * pEventLink;
 
-            break; /* If the right event is received, exit. */
+    IotMutex_Lock( &threadSafetyMutex );
+
+    /* Get the event associated to the callback */
+    IotContainers_ForEach( &eventQueueHead, pEventLink )
+    {
+        pEvent = IotLink_Container( BLEHALEventsInternals_t, pEventLink, eventList );
+
+        if( ( xEventsToCheck & pEvent->xEventTypes ) == pEvent->xEventTypes )
+        {
+            eventSet |= pEvent->xEventTypes;
         }
     }
 
     IotMutex_Unlock( &threadSafetyMutex );
 
-    return pvPtr;
+    return eventSet;
+}
+
+/* This function first check if an event is waiting in the list. If not, it will go and wait on the queue.
+ * When an event is received on the queue, if it is not the expected event, it goes on the waiting list.
+ */
+BLEHALEventsTypes_t IotTestBleHal_WaitForEvents( BLEHALEventsTypes_t xEventsToWaitFor,
+                                                 uint32_t timeoutMs )
+{
+    BLEHALEventsTypes_t eventsSet;
+
+    IotLog( IOT_LOG_DEBUG,
+            &_logHideAll,
+            "WaitForEvents: %0x",
+            xEventsToWaitFor );
+
+    eventsSet = checkQueueForEvents( xEventsToWaitFor );
+
+    /* If event is not waiting then wait for it. */
+    if( eventsSet == eBLEHALEventNone )
+    {
+        do
+        {
+            /* TODO check event list here */
+            if( IotSemaphore_TimedWait( &eventSemaphore, timeoutMs ) == true )
+            {
+                eventsSet = checkQueueForEvents( xEventsToWaitFor );
+
+                if( eventsSet != eBLEHALEventNone )
+                {
+                    break; /* If the any of the event is received, exit. */
+                }
+            }
+            else
+            {
+                /* Timeout occurred. Exit. */
+                break;
+            }
+        } while( 1 );
+    }
```

Comment:
This loop is a bit awkward to read considering there's only one looping path.

Perhaps?
```c
while(IotSemaphore_TimedWait( &eventSemaphore, timeoutMs ) == true ))
{
    if(checkQueueForEvents( xEventsToWaitFor ) != eBLEHALEventNone )
    {
        break;
    }
}
```

Reply:
Moved this to loop statement.